src/api/app/controllers/trigger_controller.rb

```diff
@@ -62,6 +63,12 @@ def set_token
   end
 
   def validate_parameters_by_token
+    # validate matching package
+    if @token.package.present?
+      raise InvalidPackage unless params[:project].nil? || @token.package.project.name == params[:project]
+      raise InvalidPackage unless params[:package].nil? || @token.package.name == params[:package]
```

Comment:
This is already happening here:

https://github.com/openSUSE/open-build-service/blob/7acb68bdb25eb10e9ee76f8c506fcbd3cf5907a1/src/api/app/controllers/concerns/triggerable.rb#L15-L21

Reply:
the test case showed that it was not catched at least without the extra check.

Reply:
the original functional misbehaviour was also that it executed the command on project level instead of specified package, since no object exists for the package.
but the "invalid" package also reached the package.

Reply:
The test case showed that it triggered `@token.package` not matter what you put into the project/package parameters...

Reply:
But I can add a note to releasenotes about the behaviour change..

Reply:
> if you specify a package it must never run on project level, that not only leads to an error in source update case.

I have no idea what you are talking about. If there is a `@token.package` nothing is *ever* executed "on project" level. What is dangerous about that?

Reply:
Even when @token.package is set, but @package is nil, you get it executed on project level.

Luckily it only leads usually to an "unknown command" error in service case, but there are also other commands which are supported also on project level.

Reply:
I added a note to the release notes.

The api documentation is unchanged as it still fits. (It is/was not documented that a registered package in the token is overwriting silently any api query parameter.

Reply:
If there is a  `@token.package` how can `@package` be nil if we do `@package = @token.package`

https://github.com/openSUSE/open-build-service/blob/7acb68bdb25eb10e9ee76f8c506fcbd3cf5907a1/src/api/app/controllers/concerns/triggerable.rb#L15

and `@package` is what we pass to `@token.call` 

https://github.com/openSUSE/open-build-service/blob/7acb68bdb25eb10e9ee76f8c506fcbd3cf5907a1/src/api/app/controllers/trigger_controller.rb#L30-L34

Reply:
because it is a remote project or a scmsync project.

Reply:
A remote or scmsync `Project` has no `Package` objects and therefor you can't assign them to a `Token`. `Token.package` is a database association. 

https://github.com/openSUSE/open-build-service/blob/4f655156314d2fb594d877a978bf2da6fd090ace/src/api/app/models/token.rb#L4